### PR TITLE
Add tests for monitoring/API

### PR DESCRIPTION
**Pull request by idoqo**
This attempts to add some tests for monitoring option (and services in general).

It involves some refactoring because previously, we had some tight couplings around e.g
- shared global configuration instance.
- API/DB coupling
- the same network name
- the same name for docker services.

Now, these can be configured to an extent, and it adds a new "service" package that can be tested independently of HTTP requests - though there's still a lot of room for improvements.

### CODE REVIEW

File: api/api_test.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+const TestAppPort = 23455
+
+func createServer(ch ClusterHandler) *http.Server {
+	addr := fmt.Sprintf(":%d", TestAppPort)
+
+	router := http.NewServeMux()
+	router.HandleFunc("/hello", Hello)
+	router.HandleFunc("/createservice", ch.CreateService)
+	router.HandleFunc("/listcluster", ch.ListCluster)
+	router.HandleFunc("/cluster", ch.GetCluster)
```
Comment (idoqo) [inline api/api_test.go:18]: This is only a subset of endpoints - but we can put everything here. Idea is to reuse this for actual integration tests.

File: api/cluster_test.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+	"go.uber.org/zap"
+	"net/http"
+	"testing"
+)
+
+// cluster tests contain unit tests for cluster-related API endpoints.
+func TestListCluster(t *testing.T) {
+	svc := &mockClusterService{}
```
Comment (idoqo) [inline api/cluster_test.go:14]: Here's a mocked service - it responds with a static list always. I'm hoping with this approach, we only need to replace it with an actual service (with actual metastore and docker client) to make this an integration test.

File: api/create.go [hunk 3]
```diff
@@ -34,155 +50,64 @@ func (c ClusterHandler) CreateService(w http.ResponseWriter, req *http.Request)
...
 	if err != nil {
-		log.Fatalf("fatal: reading from readall body %v", err)
+		c.logger.Error("error reading request body", zap.Error(err))
+		http.Error(w,"error reading request body", http.StatusInternalServerError)
+		return
 	}
-
 	err = json.Unmarshal(byteArray, &s)
-	log.Printf("%d %d %d %d", s.Version.Maj, s.Version.Min, s.Db.CPU, s.Db.Memory)
-	if s.UserID == "" && apiKeyHeader != "" {
```
Comment (viggy28) [inline api/create.go:52]: nice cleanup.

File: api/mock_cluster_service_test.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+type mockClusterService struct {
+	mock.Mock
+}
+
+// CreateService provides a mock function with given fields: ctx, info
+func (_m *mockClusterService) CreateService(ctx context.Context, info *metastore.ClusterInfo) error {
+	ret := _m.Called(ctx, info)
+
+	var r0 error
+	if rf, ok := ret.Get(0).(func(context.Context, *metastore.ClusterInfo) error); ok {
```
Comment (viggy28) [inline api/mock_cluster_service_test.go:22]: This syntax looks new to me. Perhaps I need to get familiar with it.

**Review by viggy28 — APPROVED** [review]

**Review by skant7 — APPROVED** [review]

**Pull request merged**